skmonitor/extern/native.c

```diff
@@ -0,0 +1,58 @@
+#include <ctype.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+#include <unistd.h>
+
+void* SKIP_SKMonitor_createTime(__uint64_t s, __uint64_t ns);
+void sk_global_lock();
+void sk_global_unlock();
+char* sk2c_string(char*);
+
+void* SKIP_SKMonitor_now() {
+  struct timespec now;
+  timespec_get(&now, TIME_UTC);
+  return SKIP_SKMonitor_createTime(now.tv_sec, now.tv_nsec);
+}
+
+__uint8_t SKIP_SKMonitor_checkFifo(char* skfifo) {
+  char* fifo = sk2c_string(skfifo);
+  if (access(fifo, F_OK) == 0) {
+    if (fifo != skfifo) free(fifo);
+    return 1;
+  }
+  if (fifo != skfifo) free(fifo);
+  return 0;
+}
+
+__uint8_t SKIP_SKMonitor_appendToFifo(char* skfifo, char* sktrace) {
+  char* trace = sk2c_string(sktrace);
+  char* fifo = sk2c_string(skfifo);
+  int fifo_write = open(fifo, O_WRONLY | O_APPEND | O_NONBLOCK);
```

Comment:
What are the semantics of nonblock with fifos? I think this just means you can write successfully if there's no reader. We'll still need to be careful that the reader can keep up otherwise we'll block the skdb processes, right? Or do we lose log events if the reader can't keep up and the OS pipe buffer fills?

Reply:
That works for me. I'd rather lose log events than stall.